client/network/src/service.rs

```diff
@@ -844,6 +840,34 @@ impl<B: BlockT + 'static, H: ExHashT> NetworkService<B, H> {
 		}
 	}
 
+	/// Variation of `request` which can connect to peers when necessary.
+	///
+	/// In some cases we do request data, without prior notification. For those cases this function
+	/// exists. It has no guarantees on connection staying alive, so you might still want to pass
+	/// false to the `connect` parameter and ensure liveness of connections via peer sets.
+	/// Connecting to a peer can take up to several seconds, if a peer exposes multiple
+	/// addresses, but is only reachable by some of them, for example.
+	///
+	/// Apart from exposing whether or not we want to connect, in case of a disconnected peer, this
+	/// function also takes a sender for sending the response back, instead of returning the
+	/// response directly.
+	pub fn send_request(
+		&self,
+		target: PeerId,
+		protocol: impl Into<Cow<'static, str>>,
+		request: Vec<u8>,
+		tx: oneshot::Sender<Result<Vec<u8>, RequestFailure>>,
+		connect: bool,
```

Comment:
```suggestion
```

In my opinion, if the user wants `connect` to be false, they should call `request`. If they want `connect` to be true, they should call `send_request`.

Reply:
If connect/not connect semantics end up being the only difference, then for sure. If not, I think having that simple use case function `request` and the power horse `send_request` with full flexibility makes sense. I don't think, that if I preferred a channel based API, that I also always want 'auto connect", as it has downsides as you mentioned.